readme.md
![App Logo](https://raw.githubusercontent.com/SimonDegraeve/bitbucket-semaphore-status/master/assets/logo-small.png)
Bitbucket Semaphore Status
=========================

Small app that will automatically update newly created pull requests in Bitbucket with the branch's Semaphore build status. This is inspired by [bitbucket-codeship-status](https://github.com/chesleybrown/bitbucket-codeship-status).

# Running on Heroku

First just deploy a free instance of the app on heroku using the button then just follow the steps below.

[![Deploy](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy)

1. Create an API Key in Bitbucket for your team and use your team name as the `username` and the API Key as your `password` in the next step.
1. Set `BITBUCKET_USERNAME` and `BITBUCKET_PASSWORD` ENV variables to match with the `username` and `password` above.
1. Add a `Pull Request POST` hook in Bitbucket for `Create / Edit / Merge / Decline` that points to your instance of this app. The URL should look something like this:
	`https://<YOUR_APP_NAME_ON_HEROKU>.herokuapp.com/pull-request/<SEMAPHORE_PROJECT_UUID>/<SEMAPHORE_AUTH_TOKEN>`
1. Now whenever a pull request is created, it should (almost instantly) get updated to have the [Semaphore Status Badge](https://semaphoreapp.com/docs/how-to-get-build-badge.html) in the description.

# Running Locally

Server runs on port `8000` by default, but will use the port set
on the environment variable `PORT` if set.

1. Run `npm install` for the initial setup.
1. Set `BITBUCKET_USERNAME` and `BITBUCKET_PASSWORD` ENV variables.
1. Run `npm start` to start the server.


changelog
v0.1.0:
  date: 2014-11-25
  changes:
    - Initial release


lib/helper.js
'use strict';

/*
 * Dependencies
 */
import * as request from 'co-request';
import * as chalk from 'chalk';
import {parse} from 'url';


/*
 * Internals
 */
var _Error = function(status = 500, message) {
  var error = new Error(message);
  error.status = status;
  return error;
};

var _responseErrorHandler = function(response, message) {
  if (response.statusCode >= 400) {
    throw new _Error(
      response.statusCode,
      response.headers['content-type'].contains('application/json') ? response.body.error.message || response.body.error : message
    );
  }
  return response;
};

var _retryOnError = function*(fnToCall, interval = 500, maxTry = 5) {
  var iterator = 0;
  var loop = function*() {
    try {
      return yield fnToCall();
    } catch (error) {
      iterator++;
      interval += interval;
      console.log(`Try ${iterator}/${maxTry} ${chalk.yellow(error)}`);
      if (iterator !== maxTry) {
        yield (callback) => setTimeout(callback, interval);
        yield loop;
      } else {
        throw error;
      }
    }
  };
  return yield loop();
};


/*
 * BitBucket Class
 */
export class BitBucket {
  constructor(username, password) {
    this.endpoint = 'https://api.bitbucket.org/2.0';
    this.auth = {username, password};
  }

  api() {
    return function*(query = '/user', method = 'GET', json = true) {
      return _responseErrorHandler(
        yield request({url: this.endpoint + query, auth: this.auth, method, json}),
        'Invalid BitBucket request'
      );
    }.apply(this, arguments);
  }

  isAuthenticated() {
    return function*() {
      try {
        return (yield this.api()).statusCode === 200;
      } catch (error) {}
      return false;
    }.apply(this, arguments);
  }
}


/*
 * SemaphoreBadge Class
 */
export class SemaphoreBadge {
  constructor(projectId, branchName, commitId, authToken, badgeType = 'semaphore') {
    Object.assign(this, {
      endpoint: 'https://semaphoreapp.com/api/v1',
      auth: {auth_token: authToken},
      projectId, branchName, commitId, badgeType
    });

    return function*() {
      yield this.fetchBranchId();
      yield this.fetchBuildId();
      return this.getBadge();
    }.apply(this, arguments);
  }

  api() {
    return function*(query = '/projects', method = 'GET', json = true) {
      return _responseErrorHandler(
        yield request({url: this.endpoint + query, qs: this.auth, method, json}),
        'Invalid Semaphore request'
      );
    }.apply(this, arguments);
  }

  fetchBranches() {
    return function*() {
      return (yield this.api(`/projects/${this.projectId}/branches`)).body;
    }.apply(this, arguments);
  }

  fetchHistory() {
    return function*() {
      return this.history = (yield this.api(`/projects/${this.projectId}/${this.branchId}`)).body;
    }.apply(this, arguments);
  }

  fetchBranchId() {
    return function*() {
      return yield _retryOnError(function*() {
        this.branchId = ((yield this.fetchBranches()).filter(({name}) => name === this.branchName)[0] || {}).id
        if (typeof this.branchId === 'undefined') {
          throw new _Error(404, 'Semaphore branch not found');
        }
      }.bind(this));
    }.apply(this, arguments);
  }

  fetchBuildId() {
    return function*() {
      return yield _retryOnError(function*() {
        this.buildId = ((yield this.fetchHistory()).builds.filter(({commit}) => commit.id === this.commitId)[0] || {}).build_number;
        if (typeof this.buildId === 'undefined') {
          throw new _Error(404, 'Semaphore commit not found');
        }
      }.bind(this));
    }.apply(this, arguments);
  }

  getUrl() {
    var {href, search} = parse(this.history.branch_history_url);
    return `${href.substr(0, href.length - search.length)}`;
  }

  getBadgeUrl() {
    return `${this.getUrl()}/${this.badgeType === 'semaphore' ? 'badge.png' : 'shields_badge.svg'}`;
  }

  getBadge() {
    return `[![Semaphore Build Status](${this.getBadgeUrl()})](${this.history.branch_url}/builds/${this.buildId})`;
  }
}


lib/server.js
'use strict';
// jshint -W124, -W106, -W101

/*
 * Dependencies
 */
import * as koa from 'koa';
import * as bodyParser from 'koa-bodyparser';
import * as render from 'koa-ejs';
import * as router from 'koa-router';
import * as request from 'co-request';
import * as chalk from 'chalk';
import {join} from 'path';
import {BitBucket, SemaphoreBadge} from './helper';


/*
 * Application
 */
var app = koa();
var port = (process.env.PORT) ? process.env.PORT : 8000;
render(app, {
  root: join(__dirname, '../views'),
  layout: false,
  viewExt: 'ejs',
  cache: true,
  debug: true
});


/*
 * Middlewares
 */
app.use(bodyParser());

var {BITBUCKET_USERNAME: bbUsername, BITBUCKET_PASSWORD: bbPassword} = process.env;
app.use(function*(next) {
  this.bitbucket = new BitBucket(bbUsername, bbPassword);
  yield next;
});

app.use(router(app));


/*
 * Routes
 */
app.get('/', function*() {
  yield this.render('index', {
    bbUsername: Boolean(bbUsername),
    bbPassword: Boolean(bbPassword),
    host: this.protocol + '://' + this.get('host'),
    isAuthenticated: yield this.bitbucket.isAuthenticated()
  });
});

app.post('/pull-request/:semaphoreProjectId/:semaphoreAuthToken', function*() {
  // Extract usefull values
  var pullRequest, pR = {}, sP = {};
  ({semaphoreProjectId: sP.projectId, semaphoreAuthToken: sP.authToken} = this.params);
  try {
    ({pullrequest_created: pullRequest} = this.request.body || {});
    ({id: pR.id,
      description: pR.description,
      source: {
        commit: {hash: pR.commitId},
        branch: {name: pR.branchName},
        repository: {full_name: pR.repoName}
      }
    } = pullRequest);
  } catch (error) {
    this.throw(400, 'Invalid Bitbucket hook data');
  }
  // If it doesn't already have Semaphore status at the start of the description, let's add it
  if (!pullRequest.description.contains('[![Semaphore Build Status')) {
    // Create Semaphore badge
    var badge = yield new SemaphoreBadge(sP.projectId, pR.branchName, pR.commitId, sP.authToken);
    pullRequest.description = `${badge}\r\n\r\n${pullRequest.description}`;
    // Post badge to BitBucket
    yield this.bitbucket.api(`/repositories/${pR.repoName}/pullrequests/${pR.id}`, 'PUT', pullRequest);
    this.status = 204;
  } else {
    this.status = 204;
  }
});


/*
 * Error handler
 */
app.on('error', function(error) {
  console.log(chalk.red(error.stack));
});


/*
 * Server
 */
var server = app.listen(port, function() {
  var host = server.address().address;
  console.log('Listening at http://%s:%s', host, port);
});
